node/libs/crypto/src/bn254/mod.rs

```diff
@@ -0,0 +1,213 @@
+//! BLS signature scheme for the BN254 curve.
+
+use crate::ByteFmt;
+use anyhow::anyhow;
+use ark_bn254::{Bn254, Fr, G1Projective as G1, G2Projective as G2};
+use ark_ec::{
+    pairing::{Pairing as _, PairingOutput},
+    Group as _,
+};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+pub use error::Error;
+use num_traits::Zero as _;
+use std::{collections::HashMap, fmt::Debug, hash::Hasher};
+
+#[doc(hidden)]
+pub mod error;
+
+#[cfg(test)]
+mod tests;
+
+pub mod hash;
+mod testonly;
+
+/// Type safety wrapper around a scalar value.
+pub struct SecretKey(pub Fr);
+
+impl SecretKey {
+    /// Gets the corresponding [`PublicKey`] for this [`SecretKey`]
+    pub fn public(&self) -> PublicKey {
+        let p = G2::generator() * self.0;
+        PublicKey(p)
+    }
+
+    /// Produces a signature using this [`SecretKey`]
+    pub fn sign(&self, msg: &[u8]) -> Signature {
+        let hash_point = hash::hash_to_g1(msg);
+        let sig = hash_point * self.0;
+        Signature(sig)
+    }
+}
+
+impl ByteFmt for SecretKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        Fr::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode secret key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+/// Type safety wrapper around G2.
+#[derive(Clone, PartialEq, Eq)]
+pub struct PublicKey(pub G2);
+
+impl ByteFmt for PublicKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        G2::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode public key: {e:?}"))
```

Comment:
It requires `ark_serialize::SerializationError` to satisfy `anyhow::context::ext::StdError` trait bound, hence to use a type wrapper which implements `Error` + `Display`. 
I thought it's too cumbersome.
What do you think?

Reply:
Yeah, for errors which are not anyhow::Context-compatible we can't do better than this AFAIK.

Reply:
Sorry, didn't check whether the error implements the [`Error`](https://doc.rust-lang.org/std/error/trait.Error.html) trait; I thought it would. `ark-serialize` uses [a shim](https://github.com/arkworks-rs/std) for the standard library for no-std use cases, which can be "rerouted" to `std` if the corresponding feature is enabled:

```toml
ark-serialize = { version = "0.4.2", features = ["std"] }
```

In this case, `SerializationError` will implement `Error`, and the problem will be gone. IMO, since we rely on `std` anyway, enabling this feature is harmless.

Reply:
Fixed.